blockonomics-woocommerce.php

```diff
@@ -115,93 +141,116 @@ function add_page()
                 }
                 else
                 {
                     $message = __('Congrats ! Setup is all done', 'blockonomics-bitcoin-payments');
-                    $type = 'updated';
-                    add_settings_error('option_notice', 'option_notice', $message, $type);
+                    display_admin_message($message, 'updated');
+
+                    // Check if merchant has received any confirmed BTC payments
+                    // If there are funds, make a withdraw request. Only set temp api key to null if success
+                    if (get_total_received() > 0)
+                    {
+                        $response = $blockonomics->make_withdraw();
+                        if ($response->response_code != 200)
+                        {
+                            $message = __('Error while making withdraw: '. $response->message, 'blockonomics-bitcoin-payments');
+                            display_admin_message($message, 'error');
+                        }
+                        else
+                        {
+                            $message = __('Your funds withdraw request has been submitted. Please check your Blockonomics registered emailid for details', 'blockonomics-bitcoin-payments');
+                            display_admin_message($message, 'updated');
+                            update_option("blockonomics_temp_api_key", null);
+                        }
+                    }
+                    // No funds in account, no need to withdraw
+                    else
+                    {
+                        update_option("blockonomics_temp_api_key", null);
+                    }
                 }
             }
         }
 
-        function gen_callback($input)
+        function display_admin_message($msg, $type)
         {
-          if ($input == 1)
-          {
-            $callback_secret = sha1(openssl_random_pseudo_bytes(20));
-            update_option("blockonomics_callback_secret", $callback_secret);
-          }
+            add_settings_error('option_notice', 'option_notice', $msg, $type);
+        }
 
-          return 0;
+        function generate_secret($force_generate = false)
+        {
+            $callback_secret = get_option("blockonomics_callback_secret");
+            if (!$callback_secret || $force_generate) {
+                $callback_secret = sha1(openssl_random_pseudo_bytes(20));
+                update_option("blockonomics_callback_secret", $callback_secret);
+            }
+        }
+
+        function get_callback_url()
+        {
+            $callback_secret = get_option('blockonomics_callback_secret');
+            $callback_url = WC()->api_request_url('WC_Gateway_Blockonomics');
+            $callback_url = add_query_arg('secret', $callback_secret, $callback_url);
+            return $callback_url;
+        }
+
+        function get_total_received()
+        {
+            $blockonomics_orders = get_option('blockonomics_orders');
```

Comment:
What if there are under or over payments? Using this we only see the requested amounts